jgrapht-core/src/main/java/org/jgrapht/alg/util/NeighborCache.java

```diff
@@ -0,0 +1,187 @@
+ * which becomes inordinately expensive when performed often.
+ * 
+ * <p>
+ * The cache also keeps track of successors and predecessors for each vertex.
+ * This means that the result of the union of calling predecessorsOf(v) and successorsOf(v) is 
+ * equal to the result of calling neighborsOf(v) for a given vertex v.
+ * 
+ * @param <V> the vertex type
+ * @param <E> the edge type
+ * 
+ * @author Szabolcs Besenyei
+ * @since November 2017
+ */
+public class NeighborCache<V, E>
+    implements GraphListener<V, E>
+{
+    private Map<V, NeighborIndex.Neighbors<V>> successors = new HashMap<>();
+    private Map<V, NeighborIndex.Neighbors<V>> predecessors = new HashMap<>();
+    private Map<V, NeighborIndex.Neighbors<V>> neighbors = new HashMap<>();
+
+    private Graph<V, E> graph;
+
+    /**
+     * Constructor
+     * 
+     * @param graph the input graph
+     * @throws NullPointerException if the input graph is null
+     */
+    public NeighborCache(Graph<V, E> graph)
+    {
+        this.graph = Objects.requireNonNull(graph);
+    }
+
+    /**
+     * Returns the unique predecessors of the given vertex if it exists in the cache, otherwise it
+     * is initialized.
+     * 
+     * @param v the given vertex
+     * @return the unique predecessors of the given vertex
+     */
+    public Set<V> predecessorsOf(V v)
+    {
+        return fetch(
+            v,
+            predecessors,
+            k -> new NeighborIndex.Neighbors<>(Graphs.predecessorListOf(graph, v)));
+    }
+
+    /**
+     * Returns the unique successors of the given vertex if it exists in the cache, otherwise it is
+     * initialized.
+     * 
+     * @param v the given vertex
+     * @return the unique successors of the given vertex
+     */
+    public Set<V> successorsOf(V v)
+    {
+        return fetch(
+            v,
+            successors,
+            k -> new NeighborIndex.Neighbors<>(Graphs.successorListOf(graph, v)));
+    }
+
+    /**
+     * Returns the unique neighbors of the given vertex if it exists in the cache, otherwise it is
+     * initialized.
+     * 
+     * @param v the given vertex
+     * @return the unique neighbors of the given vertex
+     */
+    public Set<V> neighborsOf(V v)
+    {
+        return fetch(
+            v,
+            neighbors,
+            k -> new NeighborIndex.Neighbors<>(Graphs.neighborListOf(graph, v)));
+    }
+
+    private Set<V> fetch(V vertex,
+        Map<V, NeighborIndex.Neighbors<V>> map,
+        Function<V, NeighborIndex.Neighbors<V>> func)
+    {
+        return map.computeIfAbsent(vertex, func).getNeighbors();
+    }
+
+    @Override
+    public void edgeAdded(GraphEdgeChangeEvent<V, E> e)
+    {
+        E edge = e.getEdge();
+        V source = graph.getEdgeSource(edge);
+        V target = graph.getEdgeTarget(edge);
+
+        if (successors.containsKey(source)) {
+            successors.get(source).addNeighbor(target);
+        } else {
```

Comment:
I would remove the else{} clauses in these if statements. Especially for large graphs, this only adds additional computational overhead, and increases the memory requirements. You can just as well rely on the successorsOf(.)/predecessorOf(.)/... functions to delay the neighbor computations until they are really needed.